File: 2025/01 January/15 Longest Subarray with Sum K/main.cpp
// Link: https://www.geeksforgeeks.org/problems/longest-sub-array-with-sum-k0809/1

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    int longestSubarray(vector<int> &arr, int k)
    {
        unordered_map<int, int> mp;
        mp[0] = -1;
        long long sum = 0;
        int n = arr.size();
        int ans = 0;
        for (int i = 0; i < n; i++)
        {
            sum += arr[i];
            if (mp.find(sum - k) != mp.end())
            {
                ans = max(ans, i - mp[sum - k]);
            }
            if (mp.find(sum) == mp.end())
                mp[sum] = i;
        }
        return ans;
    }
};

int main()
{
    Solution object;
    vector<int> arr = {10, 5, 2, 7, 1, -10};
    int k = 15;
    cout << object.longestSubarray(arr, k) << endl;

    arr = {-5, 8, -14, 2, 4, 12};
    k = -5;
    cout << object.longestSubarray(arr, k) << endl;

    arr = {10, -10, 20, 30};
    k = 5;
    cout << object.longestSubarray(arr, k) << endl;
    return 0;
}

File: 2025/01 January/15 Longest Subarray with Sum K/main.py
# Link: https://www.geeksforgeeks.org/problems/longest-sub-array-with-sum-k0809/1


class Solution:
    def longest_subarray(self, arr, k):
        prefix_sum_map = {}
        prefix_sum_map[0] = -1
        total_sum = 0
        n = len(arr)
        ans = 0

        for i in range(n):
            total_sum += arr[i]
            if (total_sum - k) in prefix_sum_map:
                ans = max(ans, i - prefix_sum_map[total_sum - k])
            if total_sum not in prefix_sum_map:
                prefix_sum_map[total_sum] = i

        return ans


if __name__ == "__main__":
    object = Solution()
    arr = [10, 5, 2, 7, 1, -10]
    k = 15
    print(object.longest_subarray(arr, k))

    arr = [-5, 8, -14, 2, 4, 12]
    k = -5
    print(object.longest_subarray(arr, k))

    arr = [10, -10, 20, 30]
    k = 5
    print(object.longest_subarray(arr, k))
